snapshotter/pkg/dmsetup/dmsetup.go

```diff
@@ -0,0 +1,310 @@
+type DeviceInfo struct {
+	Name            string
+	BlockDeviceName string
+	TableLive       bool
+	TableInactive   bool
+	Suspended       bool
+	ReadOnly        bool
+	Major           uint32
+	Minor           uint32
+	OpenCount       uint32 // Open reference count
+	TargetCount     uint32 // Number of targets in the live table
+	EventNumber     uint32 // Last event sequence number (used by wait)
+}
+
+var (
+	errTable     map[string]unix.Errno
+	errTableInit sync.Once
+)
+
+// CreatePool creates a device with the given name, data and metadata file and block size (see "dmsetup create")
+func CreatePool(poolName, dataFile, metaFile string, blockSizeSectors uint32) error {
+	thinPool, err := makeThinPoolMapping(dataFile, metaFile, blockSizeSectors)
+	if err != nil {
+		return err
+	}
+
+	_, err = dmsetup("create", poolName, "--table", thinPool)
+	return err
+}
+
+// ReloadPool reloads existing thin-pool (see "dmsetup reload")
+func ReloadPool(deviceName, dataFile, metaFile string, blockSizeSectors uint32) error {
+	thinPool, err := makeThinPoolMapping(dataFile, metaFile, blockSizeSectors)
+	if err != nil {
+		return err
+	}
+
+	_, err = dmsetup("reload", deviceName, "--table", thinPool)
+	return err
+}
+
+// CreateDevice sends "create_thin <deviceID>" message to the given thin-pool
+func CreateDevice(poolName string, deviceID int) error {
+	_, err := dmsetup("message", poolName, "0", fmt.Sprintf("create_thin %d", deviceID))
+	return err
+}
+
+// ActivateDevice activates the given thin-device using the 'thin' target
+func ActivateDevice(poolName string, deviceName string, deviceID int, size uint64, external string) error {
+	mapping := makeThinMapping(poolName, deviceID, size, external)
+	_, err := dmsetup("create", deviceName, "--table", mapping)
+	return err
+}
+
+// SuspendDevice suspends the given device (see "dmsetup suspend")
+func SuspendDevice(deviceName string) error {
+	_, err := dmsetup("suspend", deviceName)
+	return err
+}
+
+// ResumeDevice resumes the given device (see "dmsetup resume")
+func ResumeDevice(deviceName string) error {
+	_, err := dmsetup("resume", deviceName)
+	return err
+}
+
+// Table returns the current table for the device
+func Table(deviceName string) (string, error) {
+	return dmsetup("table", deviceName)
+}
+
+// CreateSnapshot sends "create_snap" message to the given thin-pool.
+// Caller needs to suspend and resume device if it is active.
+func CreateSnapshot(poolName string, deviceID int, baseDeviceID int) error {
+	_, err := dmsetup("message", poolName, "0", fmt.Sprintf("create_snap %d %d", deviceID, baseDeviceID))
+	return err
+}
+
+// DeleteDevice sends "delete <deviceID>" message to the given thin-pool
+func DeleteDevice(poolName string, deviceID int) error {
+	_, err := dmsetup("message", poolName, "0", fmt.Sprintf("delete %d", deviceID))
+	return err
+}
+
+// RemoveDevice removes a device (see "dmsetup remove")
+func RemoveDevice(deviceName string, force, retry, deferred bool) error {
```

Comment:
I'm wondering if this might be slightly more friendly (even if it is more complicated):

```go
type RemoveDeviceOpt struct {
	Force    bool
	Retry    bool
	Deferred bool
}

type RemoveDeviceOptFn func(*RemoveDeviceOpt) error

func RemoveWithForce(opt *RemoveDeviceOpt) error {
	opt.Force = true
}

// ...

func RemoveDevice(deviceName string, opts ...RemoveDeviceOptFn) error {
	options := RemoveDeviceOpt{}
	for _, opt := range opts {
		if err := opt(options); err != nil {
			return err
		}
	}
	// ...
}
```

The reason it might be more friendly is that you could write `dmsetup.RemoveDevice("myDevice", RemoveWithForce, RemoveWithDeferred)` instead of `dmsetup.RemoveDevice("myDevice", true, false, true)`.